### PR TITLE
fix(curriculum): update debugging challenge (fcc vs browser console)

### LINKED ISSUE #16958: [BETA]: Debugging: Differences b/w FCC & Browser Console (Lesson doesn't match output, one test might be passing erroneously)
#### Challenge Name
https://beta.freecodecamp.org/en/challenges/debugging/understanding-the-differences-between-the-freecodecamp-and-browser-console

#### Issue Description
1. The lesson states (excerpting for brevity):
>When you load and run an ordinary JavaScript file in your browser the console.log() statements will print exactly what you tell them to print to the browser console the exact number of times you requested. 

>This lends itself to some interesting behavior and might trip you up in the beginning, because a logged value that you expect to see only once may print out many more times depending on the number of tests and the values being passed to those tests.

The editor text says:
```
// Open your browser console
let outputTwo = "This will print to the browser console 2 times";
// Use console.log() to print the outputTwo variable
console.log(outputTwo);
```

However, Firefox & Chrome browser consoles both exhibit the same behavior, only printing the line once (see screenshots). I specifically left out the `clear.console()` line to see the duplicated console output, but it doesn't work.

2. The tests pass whether `clear.console()` is written in the editor or not. (screenshot below)

#### Browser Information
* Browser Name, Version: Firefox 59.0.1 && Chrome 65.0.3325.181
* Operating System: Win 8.1, 64-bit
* Mobile, Desktop, or Tablet: Desktop

#### Your Code
```
// Open your browser console
let outputTwo = "This will print to the browser console 2 times";
// Use console.log() to print the outputTwo variable
console.log(outputTwo);

let outputOne = "Try to get this to log only once to the browser console";
// Use console.clear() in the next line to print the outputOne only once

// Use console.log() to print the outputOne variable
console.log(outputOne);
```
#### Screenshot
PROBLEM#1:
Firefox Console:
image

Chrome Console:
image

PROBLEM#2
image

### PR DESCRIPTION
**Pull request by moT01**
I did a quick rewrite of this challenge so it makes more sense.

- [x] I have read [freeCodeCamp's contribution guidelines](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] None of my changes are plagiarized from another source without proper attribution.
- [x] All the files I changed are in the same world language (for example: only English changes, or only Chinese changes, etc.)
- [x] My changes do not use shortened URLs or affiliate links.

Closes #16958

### CODE REVIEW

**Review by Ast2435 — APPROVED** [review]

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/debugging/understanding-the-differences-between-the-freecodecamp-and-browser-console.english.md [hunk 1]
```diff
@@ -44,17 +41,14 @@ tests:
...
-// Use console.clear() in the next line to print the outputOne only once
+let output = "Get this to log once in the browser console and twice in the freeCodeCamp console";
+// Use console.clear() on the next line to clear the browser console.
 
 
-// Use console.log() to print the outputOne variable
+// Use console.log() to print the output variable
 
 
+// Check the two consoles to see the difference. The freeCodeCamp console should have printed the variable twice, one for each test of this challenge. The browser console should only print the variable once becuase you cleared it first.
```
Comment (ojeytonwilliams) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/debugging/understanding-the-differences-between-the-freecodecamp-and-browser-console.english.md]:
```suggestion
// Check the two consoles to see the difference. The freeCodeCamp console should have printed the variable twice, once for each test of this challenge. The browser console should only print the variable once because you cleared it first.
```

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/debugging/understanding-the-differences-between-the-freecodecamp-and-browser-console.english.md [hunk 2]
```diff
@@ -68,9 +62,15 @@ let outputOne = "Try to get this to log only once to the browser console";
 
 
 ```js
-let outputTwo = "This will print to the browser console 2 times"; console.log(outputTwo); let outputOne = "Try to get this to log only once to the browser console";
+// Open your browser console
```
Comment (ojeytonwilliams) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/debugging/understanding-the-differences-between-the-freecodecamp-and-browser-console.english.md]:
```suggestion
// Open your browser console.
```

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/debugging/understanding-the-differences-between-the-freecodecamp-and-browser-console.english.md [hunk 3]
```diff
@@ -68,9 +62,15 @@ let outputOne = "Try to get this to log only once to the browser console";
...
+let output = "Get this to log once in the browser console and twice in the freeCodeCamp console";
+// Use console.clear() on the next line to clear the browser console.
 console.clear();
-console.log(outputOne);
+
+// Use console.log() to print the output variable
```
Comment (ojeytonwilliams) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/debugging/understanding-the-differences-between-the-freecodecamp-and-browser-console.english.md]:
```suggestion
// Use console.log() to print the output variable.
```

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/debugging/understanding-the-differences-between-the-freecodecamp-and-browser-console.english.md [hunk 4]
```diff
@@ -44,17 +41,14 @@ tests:
...
-
-
-let outputOne = "Try to get this to log only once to the browser console";
-// Use console.clear() in the next line to print the outputOne only once
+let output = "Get this to log once in the browser console and twice in the freeCodeCamp console";
+// Use console.clear() on the next line to clear the browser console.
 
 
-// Use console.log() to print the outputOne variable
+// Use console.log() to print the output variable
```
Comment (ojeytonwilliams) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/debugging/understanding-the-differences-between-the-freecodecamp-and-browser-console.english.md]:
```suggestion
// Use console.log() to print the output variable.
```

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/debugging/understanding-the-differences-between-the-freecodecamp-and-browser-console.english.md [hunk 5]
```diff
@@ -44,17 +41,14 @@ tests:
 
 ```js
 // Open your browser console
```
Comment (ojeytonwilliams) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/debugging/understanding-the-differences-between-the-freecodecamp-and-browser-console.english.md]:
```suggestion
// Open your browser console.
```

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
I made a few tiny nitpicks and fixed a couple of typos.  Overall it looks much better, though.

**Review by ojeytonwilliams — APPROVED** [review]

**Comment by ojeytonwilliams** [comment]
Thanks @moT01, LGTM :+1:

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/debugging/understanding-the-differences-between-the-freecodecamp-and-browser-console.english.md [hunk 6]
```diff
@@ -7,31 +7,28 @@ forumTopicId: 301193
...
 tests:
-  - text: Use <code>console.log()</code> to print the <code>outputTwo</code> variable.  In your Browser Console this should print out the value of the variable two times.
-    testString: assert(code.match(/console\.log\(outputTwo\)/g));
-  - text: Use <code>console.log()</code> to print the <code>outputOne</code> variable.
-    testString: assert(code.match(/console\.log\(outputOne\)/g));
-  - text: Use <code>console.clear()</code> to modify your output so that <code>outputOne</code> variable only outputs once.
+  - text: Use <code>console.clear()</code> to clear the browser console.
     testString: assert(code.match(/^(\s*console.clear\(\);?\s*)$/gm));
+  - text: Use <code>console.log()</code> to print the <code>output</code> variable.
+    testString: assert(code.match(/console\.log\(output\)/g));
```
Comment (RandellDawson) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/debugging/understanding-the-differences-between-the-freecodecamp-and-browser-console.english.md]: To allow for white space, can we use a helper?
```suggestion
    testString: const noSpaces = code.replace(/\s/g, ''); assert(code.match(/console\.log\(output\)/));
```

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/debugging/understanding-the-differences-between-the-freecodecamp-and-browser-console.english.md [hunk 7]
```diff
@@ -7,31 +7,28 @@ forumTopicId: 301193
...
 
 ```yml
 tests:
-  - text: Use <code>console.log()</code> to print the <code>outputTwo</code> variable.  In your Browser Console this should print out the value of the variable two times.
-    testString: assert(code.match(/console\.log\(outputTwo\)/g));
-  - text: Use <code>console.log()</code> to print the <code>outputOne</code> variable.
-    testString: assert(code.match(/console\.log\(outputOne\)/g));
-  - text: Use <code>console.clear()</code> to modify your output so that <code>outputOne</code> variable only outputs once.
+  - text: Use <code>console.clear()</code> to clear the browser console.
     testString: assert(code.match(/^(\s*console.clear\(\);?\s*)$/gm));
```
Comment (RandellDawson) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/debugging/understanding-the-differences-between-the-freecodecamp-and-browser-console.english.md]: To allow for whites space can we use a helper function?  Also, if we are making the semi-colon optional, might as well get rid of it. I left them in, but do we really need the `^` and `$` here?
```suggestion
    testString: const noSpaces = code.replace(/\s/g, ''); assert(noSpaces.match(/^console.clear\(\)$/gm));
```

**Review by RandellDawson — CHANGES_REQUESTED** [review]
See my small suggestions.

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/debugging/understanding-the-differences-between-the-freecodecamp-and-browser-console.english.md (hunk 7 above)
Comment (moT01) [reply curriculum/challenges/english/02-javascript-algorithms-and-data-structures/debugging/understanding-the-differences-between-the-freecodecamp-and-browser-console.english.md]: I didn't really take a look at the tests, just changed what was there to check for the new variable name. No, we probably don't need the `^` or `?` or the `m`.

If you want to change your suggestion, I can commit them here.

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/debugging/understanding-the-differences-between-the-freecodecamp-and-browser-console.english.md (hunk 6 above)
Comment (moT01) [reply curriculum/challenges/english/02-javascript-algorithms-and-data-structures/debugging/understanding-the-differences-between-the-freecodecamp-and-browser-console.english.md]: This should work here since we are logging variables. Don't you want it to be like this though...
```suggestion
assert(noSpaces.match(/console\.log\(output\)/));
```
`noSpaces.match` instead of `code.match` like the other suggestion?

Comment (RandellDawson) [reply curriculum/challenges/english/02-javascript-algorithms-and-data-structures/debugging/understanding-the-differences-between-the-freecodecamp-and-browser-console.english.md]: Yeah, I meant to replace the `code` variable with `noSpaces`.  I will update my suggestion so you can commit it.

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/debugging/understanding-the-differences-between-the-freecodecamp-and-browser-console.english.md (hunk 7 above)
Comment (RandellDawson) [reply curriculum/challenges/english/02-javascript-algorithms-and-data-structures/debugging/understanding-the-differences-between-the-freecodecamp-and-browser-console.english.md]: I have edited both of my original suggestions.

**Comment by moT01** [comment]
When testing this again - I noticed that the first test passes with no code entered because the code tested for is in the comment of the seed code - @RandellDawson @ojeytonwilliams Don't you guys have another helper function for removing comments? Or maybe we could just reword it to say "Clear the console" or something.

**Comment by RandellDawson** [comment]
```js
const removeJSComments = str => str.replace(/\/\*[\s\S]*?\*\/|\/\/.*$/gm, '');
```

**Comment by ojeytonwilliams** [comment]
@RandellDawson and I have been discussing making these available in all tests as they turn out to be useful so often.  I'll try and help out with this soon.  Well, soonish...

**Review by RandellDawson — APPROVED** [review]

**Comment by RandellDawson** [comment]
Thank you for your contribution to the page! 👍
We're happy to accept these changes, and look forward to future contributions. 🎉

**Comment by moT01** [comment]
Yea - They have been used in quite a few places, it would be nice if we could keep them in a global utils folder or something and import them so we don't have to rewrite them in each file.
